src/libstore/derivations.cc

```diff
@@ -1097,36 +1097,52 @@ Derivation Derivation::fromJSON(
     const Store & store,
     const nlohmann::json & json)
 {
+    using nlohmann::detail::value_t;
+
     Derivation res;
 
-    res.name = json["name"];
+    if (json.type() != value_t::object) {
+        throw Error(
+            "Expected JSON of derivation to be of type '%s', but it is of type '%s'",
+            "object", // Colour this word as well
+            json.type_name());
+    }
+
+    printInfo("Reading name");
+    res.name = valueOfTypeAt(json, value_t::string, "name");
 
     {
-        auto & outputsObj = json["outputs"];
-        for (auto & [outputName, output] : outputsObj.items()) {
+        printInfo("Reading outputs");
```

Comment:
See `addTrace` in `util.hh`. I think that is a better way to add context than this logging.

Reply:
Currently that is only used in libexpr, but it should work here too. See libexpr for how it is used.

Reply:
Ah those were actually leftover prints from debugging, but I tried out `e.addTrace` and it's pretty nice! (though the function doesn't exist in `util.hh`)

```
$ nix derivation show nixpkgs#hello | jq '.[] | .outputs = { out: "/nix/store/8j3f8j-hello" }' | nix derivation add
error:
       … while reading key 'outputs'

       error: Expected JSON value with key 'out' to be of type 'object' but it is of type 'string'

$ nix derivation show nixpkgs#hello | jq '.[] | .outputs = { out: { path: "/usr/bin/hello" } }' | nix derivation add
error:
       … while reading key 'outputs'

       error: path '/usr/bin/hello' is not in the Nix store
```

A little sad that we can't get line-numbers for the JSON values, but getting a good trace is pretty nice already, and will be even more useful when I can add context for every separate derivation.